zjit/src/hir.rs

```diff
@@ -1386,6 +1397,15 @@ impl Function {
                         let replacement = self.push_insn(block, Insn::Const { val: Const::Value(unsafe { (*ice).value }) });
                         self.make_equal_to(insn_id, replacement);
                     }
+                    Insn::ObjToString { val, call_info, cd, state, .. } => {
+                        if self.is_a(val, types::StringExact) {
```

Comment:
I think this is overly strict.

```suggestion
                        if self.is_a(val, types::String) {
```

Since the interpreter checks for T_STRING, it's noop on string subclasses as well.

cc @tekknolagi the `is_a` terminology is confusing to me, because `subclass_instance.is_a(String)`, but I think `self.is_a(val, types::StringExact)` is true only when `val.class == ::String`.

I'm not going to hold up this PR over this, though and we can fix this separately.

Reply:
I see `is_a` [here](https://github.com/ruby/ruby/blob/34eaa6418e5c5b8639add323dbfd531b32a7d4a3/zjit/src/hir.rs#L1081-L1083) where `self.is_a(val, types::StringExact)` if `self.type_of(val).is_subtype(type)`, would that cover string subcalsses? But I'm also down to go with `self.is_a(val, types::String)` here to be more clear about the intention.

Reply:
There is `types::StringSubclass` and `types::StringSubclass.is_subclass(types::StringExact) == false`. So no it's not covered.